File: framework/languages/java/samples/java/TicTacToe.SessionGateway/Server/Session/src/main/java/systems/zlink/samples/tictactoe/sessiongateway/server/session/SessionServerHostFactory.java
package systems.zlink.samples.tictactoe.sessiongateway.server.session;

import org.springframework.boot.WebApplicationType;
import org.springframework.boot.autoconfigure.SpringBootApplication;
import org.springframework.boot.builder.SpringApplicationBuilder;
import org.springframework.context.annotation.Bean;
import systems.zlink.framework.spring.EnableZLinkFramework;
import systems.zlink.framework.spring.ZLinkFrameworkOptionsCustomizer;
import systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory;
import systems.zlink.samples.tictactoe.sessiongateway.server.session.sessions.handlers.PlayNotificationRelay;
import systems.zlink.samples.tictactoe.sessiongateway.shared.configuration.SampleTopology;
import systems.zlink.samples.tictactoe.sessiongateway.shared.configuration.SampleTopology.SampleSessionNode;



@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = SessionServer.class)
public final class SessionServerHostFactory {
    private SessionServerHostFactory() {
    }

    public static AutoCloseable start(String... args) {
        SpringApplicationBuilder builder = new SpringApplicationBuilder(SessionServerHostFactory.class)
            .web(WebApplicationType.NONE)
            .initializers(context ->
                context.getBeanFactory().registerSingleton(
                    "sampleSessionNode",
                    sampleSessionNode(args)));
        builder.application().setKeepAlive(true);
        return builder.run(args)::close;
    }

    @Bean
    ZLinkFrameworkOptionsCustomizer sessionOptions(SampleSessionNode sessionNode) {
        return options -> {
            options.useDiscovery(discovery ->
                discovery.addRegistryEndpoint(SampleTopology.RegistryRouterEndpoint));
            SessionServer.configureRelayNode(options, sessionNode);
            SessionServer.configure(options, sessionNode);
        };
    }

    private static SampleSessionNode sampleSessionNode(String[] args) {
        for (String arg : args) {
            if ("--reconnect".equals(arg)) {
                return SampleTopology.reconnectSession();
            }
        }
        return SampleTopology.primarySession();
    }

    @Bean
    PlayerSessionDirectory playerSessionDirectory() {
        return new PlayerSessionDirectory();
    }

    @Bean
    PlayNotificationRelay playNotificationRelay(PlayerSessionDirectory sessions) {
        return new PlayNotificationRelay(sessions);
    }
}

File: framework/languages/java/samples/kotlin/TicTacToe.SessionGateway/Server/Session/src/main/kotlin/systems/zlink/samples/kotlin/tictactoe/sessiongateway/server/session/SessionServerHostFactory.kt
package systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session

import org.springframework.boot.WebApplicationType
import org.springframework.boot.autoconfigure.SpringBootApplication
import org.springframework.boot.builder.SpringApplicationBuilder
import org.springframework.context.ApplicationContextInitializer
import org.springframework.context.ConfigurableApplicationContext
import org.springframework.context.annotation.Bean
import systems.zlink.framework.spring.EnableZLinkFramework
import systems.zlink.framework.spring.ZLinkFrameworkOptionsCustomizer
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.PlayerSessionDirectory
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.server.session.sessions.handlers.PlayNotificationRelay
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.configuration.SampleSessionNode
import systems.zlink.samples.kotlin.tictactoe.sessiongateway.shared.configuration.SampleTopology



@EnableZLinkFramework
@SpringBootApplication(
    proxyBeanMethods = false,
    scanBasePackageClasses = [SessionServer::class],
)
class SessionServerHostFactory {
    @Bean
    fun sessionOptions(sessionNode: SampleSessionNode): ZLinkFrameworkOptionsCustomizer =
        ZLinkFrameworkOptionsCustomizer { options ->
            options.useDiscovery { discovery ->
                discovery.addRegistryEndpoint(SampleTopology.RegistryRouterEndpoint)
            }
            SessionServer.configureRelayNode(options, sessionNode)
            SessionServer.configure(options, sessionNode)
        }

    @Bean
    fun playerSessionDirectory(): PlayerSessionDirectory =
        PlayerSessionDirectory()

    @Bean
    fun playNotificationRelay(sessions: PlayerSessionDirectory): PlayNotificationRelay =
        PlayNotificationRelay(sessions)

    companion object {
        fun start(args: Array<String> = emptyArray()): AutoCloseable {
            val builder = SpringApplicationBuilder(SessionServerHostFactory::class.java)
                .web(WebApplicationType.NONE)
                .initializers(ApplicationContextInitializer<ConfigurableApplicationContext> { context ->
                    context.beanFactory.registerSingleton(
                        "sampleSessionNode",
                        sampleSessionNode(args),
                    )
                })
            builder.application().setKeepAlive(true)
            val context = builder.run(*args)
            return AutoCloseable { context.close() }
        }

        private fun sampleSessionNode(args: Array<String>): SampleSessionNode =
            if (args.any { it == "--reconnect" }) {
                SampleTopology.reconnectSession()
            } else {
                SampleTopology.primarySession()
            }
    }
}
